### ISSUE TITLE
Cannot access MySQL through agama running in docker

### ISSUE THREAD

**Issue by MrLightWood** [romankuchin/ica0002-2020#54]
Good day Mr.Juri and Mr.Roman

I was doing lab 12 and encountered one problem with MySQL. My Agama service is running in docker and is accessible by my link http://193.40.156.86:9380/

As it can be seen MySQL service is not found. However, my MySQL service is working properly. I guess it is because of env variable we provide to docker agama. Can docker access my 2nd VM machine through docker running on VM 1?

Here is my code
![image](https://user-images.githubusercontent.com/17985968/99564665-64bedb80-29d3-11eb-96d9-1e5b6ba5d706.png)

**Comment by romankuchin** [comment]
Hi,

you're using "localhost" as mysql_host, but the problem is that localhost inside container it is container itself. You don't have MySQL running inside "agama" container.

Can be tested as follows:
Test MySQL connection from VM:
```
root@mrlightwood-1:~# nc -v localhost 3306
Connection to localhost 3306 port [tcp/mysql] succeeded!
_
5.7.32-0ubuntu0.18.04.1-log
                           ]*y~rA���t+F6z3cioQ]mysql_native_password
```
Test the same from container:
```
root@mrlightwood-1:~# docker exec 29f67bce22ea nc -v 127.0.0.1 3306
<no output>
```
Testing with not localhost from container:
```
root@mrlightwood-1:~# docker exec 29f67bce22ea nc -v 192.168.42.93 3306
192.168.42.93 (192.168.42.93:3306) open
_
5.7.32-0ubuntu0.18.04.1-log
                           [*`{DH#���ki
FZ%[l^mysql_native_password
```
You can use docker bridge IP as well:
```
root@mrlightwood-1:~# ip a show dev docker0
3: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:c1:83:03:95 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
root@mrlightwood-1:~# docker exec 29f67bce22ea nc -v 172.17.0.1 3306
172.17.0.1 (172.17.0.1:3306) open
_
5.7.32-0ubuntu0.18.04.1-logZhR*<%R���
                                     EX47hoImysql_native_password
```

Welcome to the wonderful world of modern networking!

**Comment by Narror** [comment]
Hello,
Isn't pymysql rather missing in the Dockerfile?

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1948, in full_dispatch_request
    rv = self.preprocess_request()
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 2242, in preprocess_request
    rv = func()
  File "/agama/agama.py", line 44, in before_request
    if not Item.metadata.tables[Item.__tablename__].exists(db.get_engine(app)):
  File "/usr/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 962, in get_engine
    return connector.get_engine()
  File "/usr/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 556, in get_engine
    self._engine = rv = self._sa.create_engine(sa_url, options)
  File "/usr/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 972, in create_engine
    return sqlalchemy.create_engine(sa_url, **engine_opts)
  File "/usr/lib/python3.8/site-packages/sqlalchemy/engine/__init__.py", line 500, in create_engine
    return strategy.create(*args, **kwargs)
  File "/usr/lib/python3.8/site-packages/sqlalchemy/engine/strategies.py", line 87, in create
    dbapi = dialect_cls.dbapi(**dbapi_args)
  File "/usr/lib/python3.8/site-packages/sqlalchemy/dialects/mysql/mysqldb.py", line 118, in dbapi
    return __import__("MySQLdb")
ModuleNotFoundError: No module named 'MySQLdb'
```

**Comment by hudolejev** [comment]
>  Isn't pymysql rather missing in the Dockerfile?

That's very possible. Will test.

**Comment by MrLightWood** [comment]
> > Isn't pymysql rather missing in the Dockerfile?
> 
> That's very possible. Will test.

MySQL still not working :(
I can ping my MySQL service, but agama still shows an internal error

**Comment by romankuchin** [comment]
Are you using the latest Dockerfile with fixes from latest PR? https://github.com/hudolejev/agama/commit/17016daed48207d1e5bdcb221d1270c7980c538e

**Comment by romankuchin** [comment]
Ah, you issue was that you used 127.0.0.1 to access MySQL. Was it fixed?

**Comment by MrLightWood** [comment]
> Ah, you issue was that you used 127.0.0.1 to access MySQL. Was it fixed?
Yep. I have changed mysql_host to my MySQL server IP and used the last version of Dockerfile. However, my Agama App still does not work
![image](https://user-images.githubusercontent.com/17985968/100763156-79e93080-33fd-11eb-85fc-b5e087a9d6d8.png)
![image](https://user-images.githubusercontent.com/17985968/100763536-f0862e00-33fd-11eb-8157-b5be7e438ebf.png)
![image](https://user-images.githubusercontent.com/17985968/100763654-21fef980-33fe-11eb-927e-8e64148b8408.png)

**Comment by romankuchin** [comment]
What's the output of "docker logs agama"?

**Comment by MrLightWood** [comment]
> What's the output of "docker logs agama"?

![image](https://user-images.githubusercontent.com/17985968/100774218-8627ba80-340a-11eb-8dc4-32d85a7ed5d6.png)

**Comment by romankuchin** [comment]
That was fixed in latest Dockerfile

**Comment by MrLightWood** [comment]
> That was fixed in latest Dockerfile

I reinstalled docker and it WORKS now. Not sure what the problem was. Sometimes the agama container image installation might give some error. I ran the playbook several times and on 3rd attempt, it worked.

Here is a screenshot of an error that appeared 2 times just in case. I will close this issue. Thanks :D
![image](https://user-images.githubusercontent.com/17985968/100780167-1289ab80-3412-11eb-9e21-d6d5148ee073.png)
